tools/benchmark/src/Runner.ts

```diff
@@ -132,63 +107,16 @@ export function benchmark(args: BenchmarkArguments): Test {
 
 		// Create and run a benchmark if we are in perfMode, else run the passed in function normally
 		if (isInPerformanceTestingMode) {
-			await options.before();
-
-			const benchmarkOptions: Benchmark.Options = {
-				maxTime: options.maxBenchmarkDurationSeconds,
-				minSamples: options.minSampleCount,
-				minTime: options.minSampleDurationSeconds,
-				defer: isAsync,
-			};
-
-			const benchmarkFunction: (deferred: {
-				resolve: Mocha.Done;
-			}) => void | Promise<unknown> = isAsync
-				? async (deferred: { resolve: Mocha.Done }) => {
-						// We have to do a little translation because the Benchmark library expects callback-based
-						// asynchronicity.
-						await argsBenchmarkFn();
-						deferred.resolve();
-				  }
-				: argsBenchmarkFn;
-
-			await new Promise<void>((resolve) => {
-				const benchmarkInstance = new Benchmark(
-					args.title,
-					benchmarkFunction,
-					benchmarkOptions,
-				);
-				// Run a garbage collection, if possible, before the test.
-				// This helps noise from allocations before the test (ex: from previous tests or startup) from
-				// impacting the test.
-				benchmarkInstance.on("start end", () => global?.gc?.());
-				benchmarkInstance.on("cycle", options.onCycle);
-				benchmarkInstance.on("complete", async () => {
-					const stats: BenchmarkData = {
-						aborted: benchmarkInstance.aborted,
-						count: benchmarkInstance.count,
-						cycles: benchmarkInstance.cycles,
-						error: benchmarkInstance.error,
-						hz: benchmarkInstance.hz,
-						stats: benchmarkInstance.stats,
-						times: benchmarkInstance.times,
-					};
-
-					test.emit("benchmark end", stats);
-
-					await options.after();
-					resolve();
-				});
-				benchmarkInstance.run();
-			});
-			return;
+			const stats = await runBenchmark(args);
+			test.emit("benchmark end", stats);
+		} else {
+			const { benchmarkFn: argsBenchmarkFn } = validateBenchmarkArguments(args);
+			await args.before?.();
+			await argsBenchmarkFn();
+			args.beforeEachBatch?.();
+			await args.after?.();
```

Comment:
Shouldn't `args.beforeEachBatch` be called before `argsBenchmarkFn`? It's the non-perf-mode so arguably batches don't make sense here, but if we're going to call it even for a single iteration in case the test definition does something relevant there, I think these are flipped.